---
title: 'ุงูููู ุงูุซุงูู: ุชุญููู ุงูุจูุงูุงุช ุงูุฌูููุฉ ูุฑุจุทูุง ุจุงูุจูุงูุงุช ุงูุฅูููููููุฉ '
output:
  html_document:
    lang: ar
    css: ../sc/style.css
    df_print: paged
  pdf_document: default
editor_options:
  markdown:
    wrap: sentence
---

--

# ูุงูู ูููุงุช MAFุ

<br><br>

ููู **MAF** ูู ุงุฎุชุตุงุฑ ูู **Mutation Annotation Format**ุ ููู ุตูุบุฉ ูููุงุช ููุงุณูุฉ ุชูุณุชุฎุฏู ุจุดูู ูุงุณุน ูู ุงููุนูููุงุชูุฉ ุงูุญูููุฉ (Bioinformatics) ูุชุฎุฒูู ูุชูุซูู ุจูุงูุงุช ุงูุทูุฑุงุช ุงูุฌูููุฉ ุงูููุชุดูุฉ ูู ุงูุฏุฑุงุณุงุช ุงูุฌูููููุฉุ ุฎุตูุตูุง ูู ูุฌุงู ุฃุจุญุงุซ ุงูุณุฑุทุงู.

ููุนุชุจุฑ ููู ุงูู MAF ุชูุณูููุง ุฌุฏููููุง (ุฌุฏูู ูุตู ููุตูู ุจููุงุตู ุฃู ุชุจููุจุงุช) ูุถู ูุนูููุงุช ูุตููุฉ ุนู ูู ุทูุฑุฉ ุฌูููุฉุ ุจุญูุซ ูุชูุญ ููุจุงุญุซูู ุชุจุงุฏู ุงูุจูุงูุงุช ูุชุญููููุง ุจุทุฑููุฉ ููุญูุฏุฉ ูููุธูุฉ.

**๐ ุงููุญุชููุงุช ุงูุฃุณุงุณูุฉ ูููู MAF:**

โช๏ธ **Gene Symbol (ุฑูุฒ ุงูุฌูู)**: ุงุณู ุงูุฌูู ุงูุฐู ุญุฏุซุช ููู ุงูุทูุฑุฉ.

โช๏ธ **Chromosome (ุงููุฑูููุณูู)**: ูููุน ุงูุทูุฑุฉ ุนูู ุงููุฑูููุณูู.

โช๏ธ **Start/End Position (ููุถุน ุงูุจุฏุงูุฉ ูุงูููุงูุฉ):** ุงููููุน ุงูุฌููููู ููุทูุฑุงุช.

โช๏ธ **Variant Classification (ุชุตููู ุงูุทูุฑุฉ):** ูุญุฏุฏ ููุน ุงูุทูุฑุฉ (ูุซู: Silentุ Missenseุ Nonsenseุ Frame-shift...).

โช๏ธ **Variant Type (ููุน ุงูุชุบูุฑ):** ูุซู SNP (ุชุนุฏุฏ ุฃุดูุงู ุงูููููููุชูุฏุงุช ุงูููุฑุฏุฉ) ุฃู Insertion ุฃู Deletion.

โช๏ธ **Reference & Tumor Allele (ุงูุฃููู ุงููุฑุฌุนู ูุฃููู ุงููุฑู):** ููุงุฑูุฉ ุจูู ุงูุชุณูุณู ุงููุฑุฌุนู ูุงูุชุบูุฑ ุงููุฑุตูุฏ.

โช๏ธ **Sample Information (ูุนูููุงุช ุงูุนููุฉ):** ูุซู ุงููุฑูุถุ ููุน ุงููุณูุฌ (ูุฑูู ุฃู ุทุจูุนู).

ููุทูู ุงูุขู ููุนูู ุนูู ูุฐุง ุงูููู!

<br><br>

------------------------------------------------------------------------

## 1. ุณูููู ุจุชููุฆุฉ ุจูุฆุฉ ุงูุนููุ ุซู ูุญููู ุงูุญุฒู ุงูุจุฑูุฌูุฉ ุงูุชู ุณูุญุชุงุฌ ุฅูููุง ุฃุซูุงุก ูุฐู ุงููุฑุดุฉ.

<br><br>

```{r Setting up the environment}

library(maftools)
library(dplyr)
library(skimr)
library(tidyverse)
library(ggplot2)
```
<br><br>

## 2. ูููู ุจูุฑุงุกุฉ ููู ุงูุจูุงูุงุช ุงููุตููุฉ ุงูุฐู ุชู ุญูุธู ูู ุงูููู ุงูุฃูู

<br><br>


``` {r load metadata}

url <- "https://raw.githubusercontent.com/Dalah-Info/health-info-bioinformatics-r/refs/heads/main/Day1/Breast_tumor_metadata_cleaned.csv"
Breast_tumor_metadata = read.csv(url,  row.names = NULL)

```

<br><br>


## 3. ูููู ุจูุฑุงุกุฉ ููู ุงูุทูุฑุงุช ุงูุฌูููุฉ: MFA

<br><br>


``` {r load maf file}

url2 <- "https://github.com/Dalah-Info/health-info-bioinformatics-r/raw/refs/heads/main/maf_file/filtered_maf_breast_only_maftools.maf.gz"
maf_file = read.maf(maf = url2)

```
<br><br>

## 4. ูุณุชูุดู ุดูู ููู MAF ูุงูุจูุงูุงุช ุงููุณุฌูุฉ ููู

<br><br>


``` {r explore maf file}

head(maf_file@data) #ุงูุจูุงูุงุช ุงููุตููุฉ ูุฌุฏูุง ูู ูุฐู ุงูุฎุงูุฉ


```

ูููู ุจุงุณุชุฎุฏุงู ุฃุฏุงุฉ **skimr** ูุชูุฑุฃ ููุง ุงูููู ูุงููุง

``` {r skim maf file}

skim(maf_file@data) 

```

<br><br>

## 5. ุฏูุฌ ููู ุงูุจูุงูุงุช ุงููุตููุฉ ูุงูุงูููููููุฉ ูุน ููู MAF

<br><br>

**ููุงุฐุงุ ๐ด**

ุนุงุฏุฉ ูุง ูุฃุชู ููู MAF ูููุตูุง ุนู ุงูููู ุงููุตููุ ููุญุชุงุฌ ูุฏูุฌููุง ูุนูุง ุญุชู ูุชููู ูู ุชุญููู ุงูุทูุฑุงุช ูุฑุจุทูุง ุจูุฐู ุงูุจูุงูุงุช.

**ูููุ ๐ด**

ูุญุชุงุฌ ุฅูู ุนููุฏ ูุงุญุฏ ูุดุชุฑู ุจูู ุงููููููุ ุนุงุฏุฉ ูููู ุฑูุฒ ุงูุนููุฉ


**ุฃูู ูุฌุฏ ุฑูุฒ ุงูุนููุฉุ ๐ด**

ูู ููู MAFุ ููุงู ููุทุงู ูู ุชุฑููุฒ ุงูุนููุงุช (Sample Barcoding) ูุฌุจ ุงูุงูุชุจุงู ูููุง:

1) ููุท TCGA

2) ููุท UUID

ุงุฎุชุฑ ุงูููุท ุงูููุฌูุฏ ูู ููู ุงูู MAF ูุฃูุถูุง ูู ุงูุจูุงูุงุช ุงููุตููุฉ (Metadata)ุ ุญุชู ุชุชููู ูู ุฑุจุท ุงูุฌุฏูููู ุจุงุณุชุฎุฏุงู ูุฐุง ุงูุนููุฏ.


``` {r patient ID in MAF file}

head(unique(maf_file@data$Tumor_Sample_UUID), 5) #ูุงููุฌุฏ ูุฏุฎูุงุช

head(unique(maf_file@data$Tumor_Sample_Barcode),5)

```
<br><br>

ุณูู ูุฎุชุงุฑ ุฑูุฒ TCGA ูุฃู UUID ุบูุฑ ููุฌูุฏ ูู ููู ุงูู MAF.

**๐ด ุฑูุฒ ุงูุนููุฉ ูู ููู MAF:**

*TCGA-OL-A66K-01A-11D-A29N-09* ูู ุฑูุฒ ูุงูู ููุนููุฉ (sample/aliquot barcode).
ููุชูููู ูู:

โช๏ธ TCGA โ ุงููุดุฑูุน (Project)

โช๏ธ OL โ ูููุน ูุตุฏุฑ ุงููุณูุฌ (Tissue Source Site - TSS)

โช๏ธ A66K โ ูุนุฑูู ุงููุดุงุฑู (ุงููุฑูุถ) (Participant/Patient ID)

โช๏ธ 01A โ ููุน ุงูุนููุฉ + ุงููุงุฑูุฑุฉ (01 = ูุฑู ุฃููู "Primary tumor" ุ A = ูุงุฑูุฑุฉ)

โช๏ธ 11D โ ุงูุฌุฒุก + ุงููุญูู ุงูุญููู (11 = ุฌุฒุกุ D = DNA)

โช๏ธ A29N โ ุงูุตููุญุฉ (Plate)

โช๏ธ 09 โ ุงููุฑูุฒ (Center)


ูู ุชูุฌุฏ ูุฐู ุงูุจูุงูุงุช ูู ููู ุงูุจูุงูุงุช ุงููุตููุฉุ ููุชุญูู ูู ุฐูู

<br><br>

``` {r explore patient ID in metadata}

head(unique(Breast_tumor_metadata$bcr_patient_barcode),5)

```
<br><br>

**๐ดุฑูุฒ ุงูุนููุฉ ูู ุงูุจูุงูุงุช ุงููุตููุฉ (Metadata):**

ูุฏููุง ููุท ุฑูุฒ ุงููุฑูุถ (Patient Barcode) ูุซู:
*TCGA-EW-A6S9*ุ ููู ูุชููู ููุท ูู ุฃูู 12 ุญุฑููุง ูู ุฑูุฒ ุงูู TCGA ุงููุงูู ุงูููุฌูุฏ ูู ููู MAF.

โ ุฅุฐู ุงููุงุนุฏุฉ ูู:
**ุฑูุฒ ุงููุฑูุถ ูู ุงูููู ุงููุตูู = ุฃูู 12 ุญุฑู ูู ุฑูุฒ ุงูุนููุฉ ูู ููู MAF**.

<br><br>

```{r get unique patient ID}

#ููุดุฆ ููู ุจูุงูุงุช ูุญุชูู ุนูู ุงูุฑูุฒ ุงูุทููู ูุงููุตูุฑ ูููุฑูุถุ ููุณุชุฎุฏูู ูุงุญูุง ููุฏูุฌ.

id_map <- data.frame(
  short_id = substr(maf_file@data$Tumor_Sample_Barcode, 1, 12),             # ูุนุฑูู ุงููุฑูุถ (12 ุญุฑู)
  Tumor_Sample_Barcode = as.character(maf_file@data$Tumor_Sample_Barcode),  # ุงููุนุฑูู ุงููุงูู ููุนููุฉ
  stringsAsFactors = FALSE
)

#ุฅุฒุงูุฉ ุงูุชูุฑุงุฑ ุนูู ูุณุชูู ุงููุฑูุถ 
id_map <- id_map[!duplicated(id_map$short_id), ]


matched_clinical_data <- merge(
  x = Breast_tumor_metadata,     # ุงูุจูุงูุงุช ุงูุณุฑูุฑูุฉ
  y = id_map,                    # ุฎุฑูุทุฉ ุงููุนุฑููุงุช ูู ููู ุงูู MAF
  by.x = "bcr_patient_barcode",  # ุงุณู ุนููุฏ ุงููุฑูุถ ูู ุงูุจูุงูุงุช ุงูุณุฑูุฑูุฉ
  by.y = "short_id",             # ุงุณู ุนููุฏ ุงููุฑูุถ (12 ุญุฑู) ูู ุงูุฎุฑูุทุฉ
  all.x = FALSE                  # ุงุญุชูุธ ููุท ุจุงููุฑุถู ุงูููุฌูุฏูู ูู ุงูู MAF
)


## ุงูุขู: matched_clinical_data ูุญูู ููุท ุงููุฑุถู ุงููุทุงุจูููุ ูุจุฏุงุฎูู ุนููุฏ Tumor_Sample_Barcode ุงููุดุชุฑู.


```

<br><br>

**โ ูุณุชุทูุน ุงูุขู ุฅูุดุงุก ููู MAF ููุชููุ ูุญุชูู ุนูู ุงูุจูุงูุงุช ุงููุตููุฉ ุงูุฅูููููููุฉ:**

<br><br>


```{r integrate maf and metadata file}

maf_file <- read.maf(maf = url2, clinicalData = matched_clinical_data, verbose = F)

```

## 5. ููุทูู ุงูุขู ูุงุณุชูุดุงู ุงูุทูุฑุงุช ุงูุฌูููุฉ ุงูููุณุฌูุฉ ูู ููู MAF!

### 5.1 ููุฎุต ุงูุทูุฑุงุช ุงูููุณุฌูุฉ ูู ููู MAFg ููู ูุฑูุถ.

```{r get sample summary}

getSampleSummary(maf_file)

```

```{r get gene summary}

getGeneSummary(maf_file)

```


```{r get gene summary 2}

getClinicalData(maf_file)

```


```{r plot maf summary}

plotmafSummary(maf = maf_file, rmOutlier = TRUE, addStat = 'median', dashboard = TRUE, titvRaw = FALSE)

```


```{r plot oncoplot of top 10 genes}

oncoplot(maf = maf_file, top = 10)

```

```{r plot titv summary }

laml.titv = titv(maf = maf_file, plot = FALSE, useSyn = TRUE)

#plot titv summary
plotTiTv(res = laml.titv)

```



```{r Variant Allele Frequencies}

plotVaf(maf = maf_file, vafCol = NULL, top = 20)


```


```{r somatic Interactions}

somaticInteractions(maf = maf_file, top = 25, pvalue = c(0.05, 0.1))

```



```{r Survival analysis}
library(data.table)
#1) encode dead alive as 0 and 1
maf_file@clinical.data$vital_status <- ifelse(maf_file@clinical.data$vital_status %in% c("DEAD", "1"), 1, 0)

# Numeric times
num <- function(x) suppressWarnings(as.numeric(as.character(x)))
dtd  <- num(maf_file@clinical.data$days_to_death)
dtlf <- num(maf_file@clinical.data$days_to_last_followup)

# Unified OS time: death time if dead, else last follow-up
maf_file@clinical.data$OS.time <- ifelse(!is.na(dtd), dtd, dtlf)

maf_file@clinical.data[,vital_status :=  as.numeric(vital_status)]
maf_file@clinical.data[,OS.time :=  as.numeric(OS.time)]


# Survival analysis
mafSurvival(maf = maf_file, genes = 'NEB', time = 'OS.time', Status = 'vital_status', isTCGA = FALSE)


```


```{r survival analysis - gene sets}

# A) Predict genesets associated with survival
#Using top 20 mutated genes to identify a set of genes (of size 2) to predict poor prognostic groups
prog_geneset = survGroup(maf = maf_file, top = 20, geneSetSize = 2, time = "OS.time", Status = "vital_status", verbose = FALSE)

mafSurvGroup(maf = maf_file, geneSet = c("FLG", "NEB"), time = "OS.time", Status = "vital_status")

```

```{r Clinical enrichment analysis - gender}

gender.ce = clinicalEnrichment(maf = maf_file, clinicalFeature = 'gender')


gender.ce$groupwise_comparision[p_value < 0.05]



plotEnrichmentResults(enrich_res = gender.ce, pVal = 0.05, geneFontSize = 0.5, annoFontSize = 0.6)


```

```{r Clinical enrichment analysis - radiation_therapy}

maf_file@clinical.data$radiation_therapy <- ifelse(
  maf_file@clinical.data$radiation_therapy %in% c("NO", "YES"),
  maf_file@clinical.data$radiation_therapy,
  NA
)


radiation_therapy.ce = clinicalEnrichment(maf = maf_file, clinicalFeature = 'radiation_therapy')


radiation_therapy.ce$groupwise_comparision[p_value < 0.05]



plotEnrichmentResults(enrich_res = radiation_therapy.ce, pVal = 0.001, geneFontSize = 0.5, annoFontSize = 0.6)


```
```{r plot sig genes}

library(data.table)
library(ggrepel)

# assume your table is called df
df <- radiation_therapy.ce$pairwise_comparision

# First extract numerator/denominator and compute relative frequencies
df[, c("mut1", "total1") := tstrsplit(n_mutated_Feature1, " of ", fixed=TRUE)]
df[, c("mut2", "total2") := tstrsplit(n_mutated_Feature2, " of ", fixed=TRUE)]

# Convert to numeric
df[, `:=`(mut1 = as.numeric(mut1),
          total1 = as.numeric(total1),
          mut2 = as.numeric(mut2),
          total2 = as.numeric(total2))]

# Compute relative values
df[, rel1 := mut1 / total1]
df[, rel2 := mut2 / total2]


# Take top 20 significant genes
top20 <- df[order(fdr)][1:20]


# Reshape to long format for ggplot
plotdf <- melt(top20,
               id.vars = "Hugo_Symbol",
               measure.vars = c("rel1", "rel2"),
               variable.name = "Feature",
               value.name = "Relative")

# Make Feature labels clearer
plotdf[, Feature := fifelse(Feature == "rel1", "Feature_1 (YES)", "Feature_2 (NO)")]

# Plot
ggplot(plotdf, aes(x = reorder(Hugo_Symbol, Relative), y = Relative, fill = Feature)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  coord_flip() +
  labs(x = "Gene", y = "Relative mutated fraction",
       title = "Top 20 significant genes by relative mutation frequency") +
  theme_minimal()


```

```{r plot }
top20[, diff_rel := rel1 - rel2]

ggplot(top20, aes(x = diff_rel, y = -log10(fdr), label = Hugo_Symbol)) +
  geom_point() +
  geom_text_repel() +
  theme_minimal() +
  labs(x = "Difference in relative mutation (Feature1 - Feature2)",
       y = "-log10(FDR)",
       title = "Effect size vs significance")
```


```{r mutation freq}


ggplot(top20, aes(x = diff_rel, y = -log10(fdr), size = mut1 + mut2, label = Hugo_Symbol)) +
  geom_point(alpha = 0.7) +
  geom_text_repel(data = subset(top20, fdr < 0.05 & (mut1 + mut2) >= 5)) +
  scale_size_continuous(name = "Total mutations") +
  theme_minimal() +
  labs(x = "Difference in relative mutation (Feature1 - Feature2)",
       y = "-log10(FDR)",
       title = "Effect size vs significance (highlighting robust genes)")


```

``` {r drug-gene interactions}

dgi = drugInteractions(maf = maf_file, fontSize = 0.75)

```




``` {r DGI - specific gene}

dnmt3a.dgi = drugInteractions(genes = "TP53", drugs = TRUE)

#Printing selected columns.
dnmt3a.dgi[,.(Gene, interaction_types, drug_name, drug_claim_name)]

```


``` {r Oncogenic Signaling Pathways}

pws = pathways(maf = maf_file, plotType = 'treemap')

pws

```

# Tumor heterogeneity and MATH scores

```{r Tumor heterogeneity and MATH scores}

library("mclust")

```
```{r tumor}

hetero_breast_tumor = inferHeterogeneity(maf = maf_file, tsb = 'TCGA-3C-AAAU-01A-11D-A41F-09')

print(hetero_breast_tumor$clusterMeans)

#Visualizing results
plotClusters(clusters = hetero_breast_tumor)


```



```{r signature}

library('NMF')
library("BSgenome.Hsapiens.UCSC.hg19", quietly = TRUE)

laml.tnm = trinucleotideMatrix(maf = maf_file, prefix = 'chr', add = TRUE, ref_genome = "BSgenome.Hsapiens.UCSC.hg19")
plotApobecDiff(tnm = laml.tnm, maf = maf_file, pVal = 0.2)

```

```{r estimateSignatures}

laml.sign = estimateSignatures(mat = laml.tnm, nTry = 6)

plotCophenetic(res = laml.sign)
```

```{r extract signature }


laml.sig = extractSignatures(mat = laml.tnm, n = 3)

```




```{r compare signature}

#Compate against original 30 signatures 
laml.og30.cosm = compareSignatures(nmfRes = laml.sig, sig_db = "legacy")

#Compate against updated version3 60 signatures 
laml.v3.cosm = compareSignatures(nmfRes = laml.sig, sig_db = "SBS")


```


```{r heatmap signature}

library('pheatmap')
pheatmap::pheatmap(mat = laml.og30.cosm$cosine_similarities, cluster_rows = FALSE, main = "cosine similarity against validated signatures")

```


```{r plot signature}

maftools::plotSignatures(nmfRes = laml.sig, title_size = 1.2, sig_db = "SBS")


```




